src/app/core/rest/workspace.service.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Lumeer: Modern Data Definition and Processing Platform
+ *
+ * Copyright (C) since 2017 Answer Institute, s.r.o. and/or its affiliates.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import {Injectable} from '@angular/core';
+
+import {Store} from '@ngrx/store';
+import {Observable} from 'rxjs/Observable';
+import {catchError, map, switchMap, take, tap} from 'rxjs/operators';
+import {isNullOrUndefined} from "util";
+import {AppState} from '../store/app.state';
+import {OrganizationConverter} from '../store/organizations/organization.converter';
+import {OrganizationModel} from '../store/organizations/organization.model';
+import {OrganizationsAction} from '../store/organizations/organizations.action';
+import {selectAllOrganizations} from '../store/organizations/organizations.state';
+import {ProjectConverter} from '../store/projects/project.converter';
+import {ProjectModel} from '../store/projects/project.model';
+import {ProjectsAction} from '../store/projects/projects.action';
+import {selectAllProjects} from '../store/projects/projects.state';
+import {OrganizationService, ProjectService} from './index';
+
+@Injectable()
+export class WorkspaceService {
+
+  constructor(private organizationService: OrganizationService,
+              private projectService: ProjectService,
+              private store: Store<AppState>) {
+  }
+
+  public getOrganizationFromStoreOrApi(code: string): Observable<OrganizationModel> {
+    return this.getOrganizationFromStore(code).pipe(
+      switchMap(organization => {
+        if (!isNullOrUndefined(organization)) {
+          return Observable.of(organization);
+        }
+        return this.getOrganizationFromApi(code);
+      })
+    );
+  }
+
+  private getOrganizationFromStore(code: string): Observable<OrganizationModel> {
+    return this.store.select(selectAllOrganizations).pipe(
+      map(organizations => organizations.find(org => org.code === code)),
+      take(1)
+    );
+  }
+
+  private getOrganizationFromApi(code: string): Observable<OrganizationModel> {
+    return this.organizationService.getOrganization(code).pipe(
+      map(organization => OrganizationConverter.fromDto(organization)),
+      tap(organization => this.store.dispatch(new OrganizationsAction.GetSuccess({organizations: [organization]}))),
```

Comment:
Why are we using services outside of the store? We must never do that! The line above can be replaced by a single action so everything is handled inside the store.

I know we want to handle errors here but there must be another way that does not violate basic NgRx principles.

Reply:
Look at official ngrx example-app https://github.com/ngrx/example-app/blob/master/src/app/guards/book-exists.ts